packages/forms/src/__tests__/form.test.tsx

```diff
@@ -417,4 +487,351 @@ describe('Form', () => {
+    const mockFn = jest.fn()
+
+    render(
+      <Form onSubmit={mockFn}>
+        <TextField name="textField" emptyAs={''} />
+        <TextAreaField name="textAreaField" emptyAs={''} />
+        <NumberField name="numberField" emptyAs={''} />
+        <DateField name="dateField" emptyAs={''} />
+        <SelectField name="selectField" defaultValue="" emptyAs={''}>
+          <option value="">No option selected</option>
+          <option value={1}>Option 1</option>
+          <option value={2}>Option 2</option>
+          <option value={3}>Option 3</option>
+        </SelectField>
+        <TextAreaField
+          name="jsonField"
+          defaultValue=""
+          validation={{ valueAsJSON: true }}
+          emptyAs={''}
+        />
+        <TextField name="fieldId" defaultValue="" emptyAs={''} />
+
+        <Submit>Save</Submit>
+      </Form>
+    )
+
+    fireEvent.click(screen.getByText('Save'))
+
+    await waitFor(() => expect(mockFn).toHaveBeenCalledTimes(1))
+
+    expect(mockFn).toBeCalledWith(
+      {
+        textField: '',
+        textAreaField: '',
+        numberField: '',
+        dateField: '',
+        selectField: '',
+        jsonField: '',
+        fieldId: '',
+      },
+      expect.anything() // event that triggered the onSubmit call
+    )
+  })
+
+  it('should have appropriate validation for NumberFields and DateFields with emptyAs={null}', async () => {
+    const mockFn = jest.fn()
+
+    render(
+      <Form onSubmit={mockFn}>
+        <NumberField
+          name="numberField"
+          data-testid="numberField"
+          emptyAs={null}
+          validation={{ min: 10 }}
+        />
+        <FieldError name="numberField" data-testid="numberFieldError" />
+        <DateField name="dateField" emptyAs={null} />
+        <Submit>Save</Submit>
+      </Form>
+    )
+    fireEvent.change(screen.getByTestId('numberField'), {
+      target: { value: 2 },
+    })
+    fireEvent.submit(screen.getByText('Save'))
+    await waitFor(() =>
+      expect(screen.getByTestId('numberFieldError')).toBeInTheDocument()
+    )
+
+    // The validation should catch and prevent the onSubmit from being called
+    expect(mockFn).not.toHaveBeenCalled()
+  })
+
+  it('should return a number for a textfield with valueAsNumber, regardless of emptyAs', async () => {
+    const mockFn = jest.fn()
+
+    render(
+      <Form onSubmit={mockFn}>
+        <TextField
+          name="tf1"
+          validation={{ valueAsNumber: true }}
+          defaultValue="42"
+        />
+        <TextField
+          name="tf2"
+          validation={{ valueAsNumber: true }}
+          defaultValue="42"
+          emptyAs={'undefined'}
+        />
+        <TextField
+          name="tf3"
+          validation={{ valueAsNumber: true }}
+          defaultValue="42"
+          emptyAs={null}
+        />
+        <TextField
+          name="tf4"
+          validation={{ valueAsNumber: true }}
+          defaultValue="42"
+          emptyAs={'NaN'}
```

Comment:
This is what I meant with my comment about `'NaN'`.  ☝️ 

Looking at `type EmptyAsValue = null | 'undefined' | 0 | ''` it seems like the test here is using an invalid value.

Reply:
Ah, thanks!    Changed it to the `emptyAs={0}` case and added a new specific test for the invalid `emptyAs` prop.  Please let me know if you see anything else!

Reply:
I don't know why, but this just popped up in my head...

> and added a new specific test for the invalid emptyAs prop.

Where did that test go? I don't think I see it.

Reply:
Hmmm, obviously I messed up!  I added the test in a new PR.  See #5242.

Thanks again Tobbe!